python_modules/libraries/dagster-airbyte/dagster_airbyte/managed/reconciliation.py

```diff
@@ -0,0 +1,510 @@
+            res, config_sources, existing_sources, workspace_id, dry_run
+        )
+        all_dests, dests_diff = reconcile_destinations(
+            res, config_dests, existing_dests, workspace_id, dry_run
+        )
+
+        reconcile_connections_post(
+            res,
+            config_connections,
+            all_sources,
+            all_dests,
+            workspace_id,
+            dry_run,
+        )
+
+        return ManagedElementDiff().join(sources_diff).join(dests_diff).join(connections_diff)
+
+
+def reconcile_normalization(
+    res: AirbyteResource,
+    existing_connection_id: Optional[str],
+    destination: InitializedAirbyteDestination,
+    normalization_config: Optional[bool],
+    workspace_id: str,
+) -> Optional[str]:
+    """
+    Reconciles the normalization configuration for a connection.
+
+    If normalization_config is None, then defaults to True on destinations that support normalization
+    and False on destinations that do not.
+    """
+    existing_basic_norm_op_id = None
+    if existing_connection_id:
+        operations = cast(
+            Dict[str, List[Dict[str, str]]],
+            check.not_none(
+                res.make_request(
+                    endpoint="/operations/list",
+                    data={"connectionId": existing_connection_id},
+                )
+            ),
+        )
+        existing_basic_norm_op = next(
+            (
+                operation
+                for operation in operations["operations"]
+                if is_basic_normalization_operation(operation)
+            ),
+            None,
+        )
+        existing_basic_norm_op_id = (
+            existing_basic_norm_op["operationId"] if existing_basic_norm_op else None
+        )
+
+    if normalization_config is not False:
+        if res.does_dest_support_normalization(destination.destination_definition_id, workspace_id):
+            if existing_basic_norm_op_id:
+                return existing_basic_norm_op_id
+            else:
+                return cast(
+                    Dict[str, str],
+                    check.not_none(
+                        res.make_request(
+                            endpoint="/operations/create",
+                            data={
+                                "workspaceId": workspace_id,
+                                "name": "Normalization",
+                                "operatorConfiguration": {
+                                    "operatorType": "normalization",
+                                    "normalization": {"option": "basic"},
+                                },
+                            },
+                        )
+                    ),
+                )["operationId"]
+        elif normalization_config is True:
+            raise Exception(
+                f"Destination {destination.destination.name} does not support normalization."
+            )
+
+    return None
+
+
+def reconcile_connections_pre(
```

Comment:
similar comment about deleting stuff here, but regardless of that discussion, I think the sequencing here is a bit confusing.

We use this function to generate a diff between the config state and the instance state (and delete unmanaged connections), but then a separate function (reconcile_connections_post), which also has to essentially calculate a diff itself, to apply non-deletion changes.

Would prefer if this functionality was combined into a single function (basically making a single call to /connections/list, then delegating out to functions like delete_connection or update_connection depending on the diff.

Reply:
The ordering is pretty important here, basically we need to:

- `reconcile_connections_pre`: Delete all connections which depend on sources or destinations that are being recreated or deleted. The underlying sources/destinations can't be deleted unless this happens first.
- `reconcile_sources/reconcile_destinations`: Update all sources and destinations, including deletion/recreation.
- `reconcile_connections_post`: (Re)create connections based on new sources and destinations.

I've added a few more comments which hopefully make this a bit more clear. If you can think of a better way to structure this, happy to take another look.

Reply:
got it -- didn't realize that airbyte wouldn't let you delete a source with dependent connections (makes sense)

I think this is a reasonable way to structure things